src/main/scala/tech/cryptonomic/conseil/generic/chain/DataTypes.scala

```diff
@@ -88,6 +88,41 @@ object DataTypes {
       .map(_.flatten.collect { case (false, fieldName) => InvalidAggregationFieldForType(fieldName) }.toList)
   }
 
+  /** Helper method for finding if queries does not contain filters on key fields or datetime fields */
+  private def findInvalidPredicateFilteringFields(query: Query, path: EntityPath, metadataService: MetadataService)
+    (implicit ec: ExecutionContext): Future[List[InvalidPredicateFiltering]] = {
+    metadataService.getEntities(path.up).flatMap { entitiesOpt =>
+      val limitedQueryEntity = entitiesOpt
+        .toList
+        .flatten
+        .find(_.name == path.entity)
+        .flatMap(_.limitedQuery)
+        .getOrElse(false)
+      if(limitedQueryEntity) {
+        query
+          .predicates
+          .traverse { predicate =>
+            metadataService.getTableAttributesWithoutUpdatingCache(path).map { attributesOpt =>
+              attributesOpt.flatMap { attributes =>
+                attributes.find(_.name == predicate.field)
+              }.toList
+            }
+          }.map(attributes => doesPredicateContainValidAttribute(attributes.flatten))
+      } else {
+        Future.successful(List.empty)
+      }
+    }
+  }
+
+  /** Checks if any of the attributes is valid for predicate */
+  private def doesPredicateContainValidAttribute(attributes: List[Attribute]): List[InvalidPredicateFiltering] = {
+    if(attributes.exists(attr => attr.keyType == KeyType.UniqueKey || attr.dataType == DataType.DateTime || attr.sufficientForQuery.getOrElse(false))) {
```

Comment:
I would move that logic to a method in an `Attribute` class.

Reply:
moved to Attribute class, regarding Options - @dorians correct me if I'm wrong - I think it is related to config deserialisation into case classes or it just reflect more accurate whats going on there.